### PR TITLE
Lock vehicle doors

### LINKED ISSUE #46160: Locks for vehicle doors
# Is your feature request related to a problem? Please describe.
Currently, human enemies can open the doors of a vehicle. Not a big deal normally, but when fighting hordes mixed with feral humans, who can use doors, it's quite annoying to have them simply walk up and open your armored vehicles door and hop in. And by annoying I mean lethal.

# Describe the solution you'd like
We currently have cargo locks that allow you to lock some vehicle storage containers to prevent NPCs from opening them, it would be nice to have a door lock part as well.

# Describe alternatives you've considered
Well, I could remove all the doors from my car and just weld myself into it, killdozer style. 

# Additional context
As a side note I think it's weird that enemies can just hop into fast moving vehicles if they happen to be next to a door while it's open.

### PR DESCRIPTION
**Pull request by Numbuh474**
#### Summary
Features "Make vehicle doors lockable and pickable"

#### Purpose of change

This is a contribution to #51383, addressing the ability to lock and unlock vehicle doors.
It also fixes #46160.

#### Describe the solution

Enable locking and unlocking of certain vehicle doors, adding the `LOCKABLE_DOOR` flag for valid doors and the `DOOR_LOCKING` flag for parts that allow locking and unlocking. A new item, the door lock, was given the DOOR_LOCKING flag and most metal doors and hatches were given the `LOCKABLE_DOOR` tag. A door must be a LOCKABLE_DOOR and have a part with the DOOR_LOCKING tag on the same frame to enable the feature. Modify relevant door opening code to respect locked vehicle door status for player and NPCs. Add lockpicking behavior for vehicle doors, allowing lockpick tools to pick open doors that are otherwise inaccessible.

Monsters and door-opening ferals aren't able to figure out locks, but NPCs will path through locked doors if given the option.  A new pair of behavior flags were added that allows you to make followers stop unlocking doors, or to also lock them if possible.

Behavior that protected door motors from damage if the door was closed was added to door locks and moved to the more generic `BOARD_INTERNAL` flag, which now also works with things installed in boards in case someone wants to add pipework or something. Doors and their locks can't be removed while a tile on the same square is in the "locked" state.

Door motor menu was modified to allow you to lock and unlock all closed doors if the door is lockable, and examining a tile with a lockable door now has options for modifying the lock state based on player position.

#### Describe alternatives you've considered

Doors could have been made intrinsically lockable without a corresponding DOOR_LOCKING part.
Part lock state currently has nothing to do with vehicle lock state, and these could be made to work together to do things like lock doors on vehicle spawn or trigger alarms. 
Lockpicking could have been excluded or made impossible for player-owned vehicles, as this gives "free" locked doors for training purposes.

#### Testing

Tested door opening and closing from inside and outside the vehicle, door locking and unlocking from both inside and outside the vehicle, door-opening monster behavior for closed vs. locked doors, and locking and unlocking via the door motor menu.

Tested having an NPC follow you into and out of a vehicle with a door lock while using the new rule toggles.

### CODE REVIEW

File: src/iexamine.cpp [hunk 1]
```diff
@@ -1793,8 +1793,26 @@ void iexamine::locked_object( Character &you, const tripoint &examp )
     map &here = get_map();
     item &best_prying = you.item_with_best_of_quality( qual_PRY, true );
     item &best_lockpick = you.item_with_best_of_quality( qual_LOCKPICK, true );
+    cata::optional<vpart_position> veh = here.veh_at( examp );
+    int locked_part = -1;
+    std::string part_name;
```
Comment (irwiss) [inline src/iexamine.cpp:1798]: Making this variable save whatever you're lockpicking (part/furniture/terrain name) here would probably save the 5 nested ternaries below

File: src/map.cpp [hunk 2]
```diff
@@ -4377,7 +4377,13 @@ bool map::open_door( Creature const &u, const tripoint &p, const bool inside,
 
         return true;
     } else if( const optional_vpart_position vp = veh_at( p ) ) {
-        int openable = vp->vehicle().next_part_to_open( vp->part_index(), true );
+        optional_vpart_position creature_veh = veh_at( u.pos() );
+        bool creature_outside = true;
+        if( creature_veh.has_value() ) {
+            creature_outside = &creature_veh->vehicle() != &veh_at( p )->vehicle();
+        }
```
Comment (irwiss) [inline src/map.cpp:4384]:
```suggestion
        bool creature_outside = creature_veh.has_value() &&
                                &creature_veh->vehicle() != &veh_at( p )->vehicle();
```

File: src/vehicle.cpp [hunk 3]
```diff
@@ -2788,11 +2788,56 @@ int vehicle::next_part_to_close( int p, bool outside ) const
 int vehicle::next_part_to_open( int p, bool outside ) const
 {
     std::vector<int> parts_here = parts_at_relative( parts[p].mount, true, true );
-
+    bool has_lock = false;
+    if( !parts_here.empty() ) {
+        has_lock = part_has_lock( parts_here.front() );
+    }
```
Comment (irwiss) [inline src/vehicle.cpp:2794]:
```suggestion
    const bool has_lock = !parts_here.empty() && part_has_lock( parts_here.front() );
```

**Review by irwiss — COMMENTED** [review]
Bunch more stuff (besides the suggested) can be const-ified
Clang is also likely to scream at you for vector.size() > 0 over !vector.empty()

I'm sure many people are waiting for this feature ♥

File: src/iexamine.cpp (hunk 1 above)
Comment (Numbuh474) [reply src/iexamine.cpp]: Done.

File: src/vehicle_use.cpp [hunk 4]
```diff
@@ -2212,19 +2212,21 @@ void vehicle::build_interact_menu( veh_menu &menu, const tripoint &p, bool with_
     const cata::optional<vpart_reference> vp_door_lock = vp.avail_part_with_feature( "DOOR_LOCKING" );
     if( vp_lockable_door && vp_door_lock && !vp_lockable_door->part().open ) {
         const size_t door_idx = vp_lockable_door->part_index();
-        if( !vp_lockable_door->part().locked ) {
+        if( player_inside && !vp_lockable_door->part().locked ) {
             menu.add( string_format( _( "Lock %s" ), vp_lockable_door->part().name() ) )
             .hotkey( "LOCK_DOOR" )
             .on_submit( [this, door_idx] {
                 this->lock_or_unlock( door_idx, true );
+                get_player_character().moves -= 100;
```
Comment (mqrause) [inline src/vehicle_use.cpp:2220]: Using `mod_moves` would be preferred over directly manipulating moves.

Comment (Numbuh474) [reply src/vehicle_use.cpp]: Done.

**Comment by Entity-y** [comment]
This looks great, thanks for putting this together. We're one step away from having dedicated ignition keys for cars.

Key fobs would be a great addition as well for remote starting, locking/unlocking, trunk access, and panic alarm.

**Comment by kevingranade** [comment]
What are NPC interactions with locked doors like? They should have the same options as players.

**Comment by Numbuh474** [comment]
> What are NPC interactions with locked doors like? They should have the same options as players.

NPCs don't currently have any logic to interact with locked doors, but it would make sense for them to be able to do so. I'll have to read over the npc pathing and behavior code to see how this should be implemented.

**Comment by Zireael07** [comment]
@Numbuh474 Probably something for a follow-up PR, though. The bigger the PR, the harder it is to review

**Comment by kevingranade** [comment]
No, it needs to happen in this PR. Having the player be the only agent that can interact with locks is a misfeature.

**Comment by Numbuh474** [comment]
> No, it needs to happen in this PR. Having the player be the only agent that can interact with locks is a misfeature.

Added relevant code for pathfinding and behavior toggles.

File: src/gates.cpp [hunk 5]
```diff
@@ -353,3 +354,115 @@ void doors::close_door( map &m, Creature &who, const tripoint &closep )
...
+    const optional_vpart_position const vp = m.veh_at( lockp );
+    const vehicle *const veh = &vp->vehicle();
+    const bool outside_vehicle = m.veh_at( who.pos() ) &&
+                                 &vp->vehicle() != &m.veh_at( who.pos() )->vehicle();
+    const int vpart = vp->part_index();
+    const int lockable = veh->next_part_to_lock( vpart, outside_vehicle );
+
+    return lockable >= 0 ?
+           true :
+           false;
```
Comment (Night-Pryanik) [inline src/gates.cpp:410]:
```suggestion
    return lockable >= 0;
```

File: src/gates.cpp [hunk 6]
```diff
@@ -353,3 +354,115 @@ void doors::close_door( map &m, Creature &who, const tripoint &closep )
...
+}
+
```
Comment (Night-Pryanik) [inline src/gates.cpp:412]:
```suggestion
```

File: src/gates.cpp [hunk 7]
```diff
@@ -353,3 +354,115 @@ void doors::close_door( map &m, Creature &who, const tripoint &closep )
...
+    const optional_vpart_position const vp = m.veh_at( lockp );
+    const vehicle *const veh = &vp->vehicle();
+    const bool outside_vehicle = m.veh_at( who.pos() ) &&
+                                 &vp->vehicle() != &m.veh_at( who.pos() )->vehicle();
+    const int vpart = vp->part_index();
+    const int unlockable = veh->next_part_to_unlock( vpart, outside_vehicle );
+
+    return unlockable >= 0 ?
+           true :
+           false;
```
Comment (Night-Pryanik) [inline src/gates.cpp:467]:
```suggestion
    return unlockable >= 0;
```

File: src/iexamine.cpp [hunk 8]
```diff
@@ -1793,8 +1793,26 @@ void iexamine::locked_object( Character &you, const tripoint &examp )
...
+    // Check if the locked thing is a lockable door part.
+    if( veh ) {
+        std::vector<vehicle_part *> parts_at_target = veh->vehicle().get_parts_at(
+                    examp, "LOCKABLE_DOOR", part_status_flag::available );
+        if( !parts_at_target.empty() ) {
+            locked_part = veh->vehicle().next_part_to_unlock(
+                              veh->vehicle().index_of_part( parts_at_target.front() ) );
+        }
+    }
+
```
Comment (Night-Pryanik) [inline src/iexamine.cpp:1808]:
```suggestion
```

File: src/vehicle.cpp [hunk 9]
```diff
@@ -2762,12 +2762,59 @@ int vehicle::next_part_to_close( int p, bool outside ) const
...
+    const std::vector<int> parts_here = parts_at_relative( parts[p].mount, true, true );
+    const bool has_lock = part_has_lock( p );
+    // We want forwards, since we open the outermost thing first (curtains), and then the innermost thing (door)
 
-    // We want forwards, since we open the innermost thing first (curtains), and then the innermost thing (door)
     for( const int &elem : parts_here ) {
         if( part_flag( elem, VPFLAG_OPENABLE ) && parts[ elem ].is_available() && parts[elem].open == 0 &&
-            ( !outside || !part_flag( elem, "OPENCLOSE_INSIDE" ) ) ) {
+            !( part( elem ).locked && has_lock ) &&
+            ( !outside || ! part_flag( elem, "OPENCLOSE_INSIDE" ) ) ) {
```
Comment (Night-Pryanik) [inline src/vehicle.cpp:2772]:
```suggestion
            ( !outside || !part_flag( elem, "OPENCLOSE_INSIDE" ) ) ) {
```

File: src/vehicle_use.cpp [hunk 10]
```diff
@@ -1200,6 +1253,39 @@ void vehicle::close( int part_index )
     }
 }
 
+
+/**
+ * Unlocks a lockable, closed part at the specified index. If it's a multipart, locks
```
Comment (Night-Pryanik) [inline src/vehicle_use.cpp:1258]:
```suggestion
 * Unlocks a lockable, closed part at the specified index. If it's a multipart, unlocks
```

File: src/vehicle_use.cpp [hunk 11]
```diff
@@ -1288,6 +1378,34 @@ void vehicle::open_or_close( const int part_index, const bool opening )
     coeff_air_dirty = true;
 }
 
+/**
+ * Locks or unlocks a lockable door part at the specified index based on the @locking value passed.
+ *
```
Comment (Night-Pryanik) [inline src/vehicle_use.cpp:1383]:
```suggestion
```

File: src/vehicle_use.cpp [hunk 12]
```diff
@@ -1288,6 +1378,34 @@ void vehicle::open_or_close( const int part_index, const bool opening )
...
+    const auto part_lock_or_unlock = [&]( const int parti, const bool locking ) {
+        vehicle_part &prt = parts.at( parti );
+        prt.locked = locking;
+        if( prt.is_fake ) {
+            parts.at( prt.fake_part_to ).locked = locking;
+        } else if( prt.has_fake ) {
+            parts.at( prt.fake_part_at ).locked = locking;
+        }
+    };
+    //find_lines_of_parts() doesn't return the part_index we passed, so we set it on it's own
```
Comment (Night-Pryanik) [inline src/vehicle_use.cpp:1400]:
```suggestion
    //find_lines_of_parts() doesn't return the part_index we passed, so we set it on its own
```

**Comment by Night-Pryanik** [comment]
I would like to see locked doors on some (most?) cars on spawn, as you described in "alternatives" section, but I guess that could wait for subsequent PRs.

Correct me if I'm wrong, but did you make it so doors can be locked and unlocked only from inside the vehicle?

**Comment by Numbuh474** [comment]
> I would like to see locked doors on some (most?) cars on spawn
So would I! As far as I can tell this would involve a lot of changes to car template json and some tweaks to vehicle spawning code, but as you can see this PR is getting quite large on its own.
> did you make it so doors can be locked and unlocked only from inside the vehicle?
That's correct. A remote vehicle controller could be used from outside too.

**Comment by Night-Pryanik** [comment]
> did you make it so doors can be locked and unlocked only from inside the vehicle?

> That's correct.

So, it will be impossible to manually lock all doors of my vehicle and go for a walk?

**Comment by Numbuh474** [comment]
> So, it will be impossible to manually lock all doors of my vehicle and go for a walk?
I did try making the lock action bidirectional, but it felt like it could be too easy for a player (or follower) to accidentally lock themselves out of a vehicle. Especially so if the aforementioned RVC or lockpicks were left inside a cargo space. I decided to go with the current behavior and make locking a "you must have the remote that can open this in order to lock it" sort of thing, but I could be convinced to change it.

**Comment by Numbuh474** [comment]
I should also mention that you technically can open a locked door from the outside without a remote, but it would require lockpicking tools.

**Comment by Night-Pryanik** [comment]
So, player must use lockpicks to manually unlock the doors from outside of the vehicle? I just thought that player will somehow get (craft?) some sort of keys to his own locks.

**Comment by Numbuh474** [comment]
This would probably involve making a key that can have a unique identifier, an interaction that tests a key against a lock, and an action for vehicle parts/items that copies the identifier into it.

As a new contributor I didn't feel confident implementing a whole new system like this. Locks of various kinds already exist in game but none of them have attempted to do that yet. Looking at similar "unique item" features: mounted turrets can't have weapon mods, radio signals only have one type, keycard readers have simple relations to doors, and so on. Generating IDs, a new item type, and the involved logic seemed like it could be difficult.

**Review by ghost — CHANGES_REQUESTED** [review]
I believe these were overlooked:
Edit: Well, that's not overly descriptive. This is the door_trunk and hddoor_trunk for clarity.
```
--- a/data/json/vehicleparts/vehicle_parts.json
+++ b/data/json/vehicleparts/vehicle_parts.json
@@ -2164,7 +2164,8 @@
       "NO_ROOF_NEEDED",
       "WINDOW",
       "SIMPLE_PART",
-      "DOOR"
+      "DOOR",
+      "LOCKABLE_DOOR"
     ],
     "breaks_into": "ig_vp_frame",
     "damage_reduction": { "all": 20 }
@@ -2189,7 +2190,7 @@
       "removal": { "skills": [ [ "mechanics", 2 ] ], "time": "30 m", "using": [ [ "vehicle_wrench_2", 1 ] ] },
       "repair": { "skills": [ [ "mechanics", 3 ] ], "time": "60 m", "using": [ [ "repair_welding_standard", 4 ] ] }
     },
-    "flags": [ "OBSTACLE", "OPENABLE", "MULTISQUARE", "BOARDABLE", "LOW_FINAL_AIR_DRAG", "WINDOW", "SIMPLE_PART", "DOOR" ],
+    "flags": [ "OBSTACLE", "OPENABLE", "MULTISQUARE", "BOARDABLE", "LOW_FINAL_AIR_DRAG", "WINDOW", "SIMPLE_PART", "DOOR", "LOCKABLE_DOOR" ],
     "breaks_into": "ig_vp_hdframe",
     "damage_reduction": { "all": 30 }
   },
```

**Comment by intergalacticdata** [comment]
I've added locks to all vehicle doors in vanilla currently. I can make those changes available to this PR or as a separate one w/ this as a dependency?

**Comment by Numbuh474** [comment]
> I've added locks to all vehicle doors in vanilla currently. I can make those changes available to this PR or as a separate one w/ this as a dependency?

I think it would be better to have a separate one since this one is getting pretty large as it is. Thanks for the contribution!

**Comment by Numbuh474** [comment]
To whom it may concern:

Whatever I just did made Github very unhappy. I apparently merged it wrong and it's reading all the changes made since the last sync with upstream master. I'll come back to this later but I suspect it might take a while to fix.

**Comment by Zireael07** [comment]
It's highly likely that you'll have to re-do the PR in this case :(

**Comment by Numbuh474** [comment]
I was afraid you were going to say that. I'll see about doing a remake and closing this one when finished.

**Comment by LyleSY** [comment]
Sorry, been there. Thanks for plunging through, it's an exciting project

File: src/vehicle_use.cpp [hunk 13]
```diff
@@ -184,11 +218,26 @@ void vehicle::control_doors()
         menu.add( _( "Close all curtains and doors" ) )
         .hotkey_auto()
         .location( get_player_character().pos() )
-        .on_submit( [&open_or_close_all] { open_or_close_all( false, "" ); } );
+        .on_submit( [&open_or_close_all, &lock_or_unlock_all] {
+            open_or_close_all( false, "" );
```
[the remark this answers is not in the document]
Comment (Numbuh474) [reply src/vehicle_use.cpp]: Re-resolved!

**Comment by Numbuh474** [comment]
Note: closed automatically because whatever I did to sync to master decided to do that. Reopening, and one of these days I should really learn how to use Git properly.

**Comment by GuardianDll** [comment]
Why closed?

**Comment by Numbuh474** [comment]
The "sync fork" button on Github web has done this twice now when it didn't before, and I'm scared to try a different method because last time I did that it nuked the PR (see comments from February).

**Comment by Maleclypse** [comment]
Ok I've cleared the merge conflicts and the errors from that.  But I'm not sure what to do with the remaining errors when I attempt to test this.
```
23:08:55.844 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1505 [finalize_prototypes] veh prototype 'Sports Car' uses invalid variant 'wheel_left' for part 'door_trunk' defaulting to ''
23:08:58.243 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1505 [finalize_prototypes] veh prototype 'Sports Car' uses invalid variant 'wheel_left' for part 'door_trunk' defaulting to ''
23:08:58.260 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1505 [finalize_prototypes] veh prototype 'Fire Engine' uses invalid variant 'horizontal_2' for part 'door_trunk' defaulting to ''
[ Previous repeated 1 times ]
23:08:58.260 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1505 [finalize_prototypes] veh prototype 'Fire Engine' uses invalid variant 'horizontal_2' for part 'door_trunk' defaulting to ''
23:08:58.269 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in truck_swat
23:08:58.983 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in truck_swat
23:08:58.988 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1505 [finalize_prototypes] veh prototype 'Stagecoach' uses invalid variant 'nw' for part 'door_wood' defaulting to ''
23:08:58.988 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1505 [finalize_prototypes] veh prototype 'Stagecoach' uses invalid variant 'ne' for part 'door_wood' defaulting to ''
23:08:59.037 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in aapc-gl
23:08:59.040 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in aapc-mg
23:08:59.042 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in apc
23:08:59.044 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in apc-gl
23:08:59.090 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in armored_car
23:08:59.096 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1505 [finalize_prototypes] veh prototype 'Hippie Van' uses invalid variant 'horizontal_2' for part 'door_trunk' defaulting to ''
23:08:59.096 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1505 [finalize_prototypes] veh prototype 'Hippie Van' uses invalid variant 'horizontal_2' for part 'door_trunk' defaulting to ''
23:08:59.117 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\vehicle.cpp:1425 [install_part] installing v_curtain would make invalid vehicle: Part requires any of: door, hatch, heavy-duty hatch, heavy-duty trunk door, reinforced windshield, trunk door, windshield, or wooden door.
23:08:59.979 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1492 [finalize_prototypes] init_vehicles: 'Survivor RV' part 'v_curtain'(91) can't be installed to -2,3
23:09:00.435 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in security_van
23:09:00.439 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in prisoner_van_1
23:09:00.440 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1484 [finalize_prototypes] unknown vehicle part hddoor_opaque in prisoner_van_1
23:09:05.842 INFO : Loaded tileset: Chibi_Ultica
23:09:05.923 INFO : Loaded tileset: Chibi_Ultica
23:09:09.386 ERROR : C:\Users\colli\Documents\GitHub\Cataclysm-DDA\src\veh_type.cpp:1658 [check] vpart migration specifies invalid id 'hddoor_opaque'
```

**Comment by RenechCDDA** [comment]
Looks like some pre-made vehicles need to be updated. Not sure on the exact syntax, but here's the spots to make it easy for you.

Sports car:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/cars.json#L622-L623

Fire Engine:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/emergency.json#L192-L197

truck_swat:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/emergency.json#L720
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/emergency.json#L740

Stagecoach:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/farm.json#L305-L306

aapc-gl:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/military.json#L79

aapc-mg:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/military.json#L177

apc:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/military.json#L275

apc-gl:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/military.json#L363

armored_car:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/vans_busses.json#L72

Hippie van:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/vans_busses.json#L354-L355

Survivor RV (curtain needs to just be removed, looks like?):
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/vans_busses.json#L1322

security van:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/vans_busses.json#L2251

prisoner_van_1:
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/vans_busses.json#L2433
https://github.com/CleverRaven/Cataclysm-DDA/blob/30f23ec0a28784af6c9678d27b6364bc5431c524/data/json/vehicles/vans_busses.json#L2437

**Comment by RenechCDDA** [comment]
Found one of your problems. Some of the parts have been moved around in the last month. Here's two instances of `door_trunk` in the PR's branch.

https://github.com/CleverRaven/Cataclysm-DDA/blob/4e7329e9a2e9c547dfd7e474b43b5f7266e15245/data/json/vehicleparts/doors.json#L230-L231

https://github.com/CleverRaven/Cataclysm-DDA/blob/4e7329e9a2e9c547dfd7e474b43b5f7266e15245/data/json/vehicleparts/vehicle_parts.json#L2236-L2238

**Comment by RenechCDDA** [comment]
Same deal with door_wood.

https://github.com/CleverRaven/Cataclysm-DDA/blob/4e7329e9a2e9c547dfd7e474b43b5f7266e15245/data/json/vehicleparts/doors.json#L195-L197

https://github.com/CleverRaven/Cataclysm-DDA/blob/4e7329e9a2e9c547dfd7e474b43b5f7266e15245/data/json/vehicleparts/vehicle_parts.json#L2712-L2714

**Comment by Maleclypse** [comment]
Pushed the changes, loaded up the game created new player, world initialized with no errors then revealed map, teleported into nearest city to test vehicle doors and ctd.

**Comment by Maleclypse** [comment]
Compiled and tested this morning and it works.  NPCs can also be commanded to lock doors as they pass through.

**Comment by sparr** [comment]
Commit comment explaining the mystery `check_only` flag: https://github.com/Numbuh474/Cataclysm-DDA/commit/46af975ca55e7b1350aea60f10cc93bc440df230#r163721522

**Pull request merged**